File: python/experiments.py
import random
import time

from matrices import random_matrix, random_vector, hilbert_matrix, mat_vec
from metrics import residual_norm, relative_error
from solvers import (
    gaussian_no_pivot,
    gaussian_partial_pivot,
    lu_decomposition,
    solve_with_lu,
)
from timing import measure_time


def print_table(headers, rows):
    headers = [str(h) for h in headers]
    rows = [[str(value) for value in row] for row in rows]

    widths = [len(h) for h in headers]
    for row in rows:
        for i in range(len(row)):
            widths[i] = max(widths[i], len(row[i]))

    print(" | ".join(headers[i].ljust(widths[i]) for i in range(len(headers))))
    print("-+-".join("-" * w for w in widths))
    for row in rows:
        print(" | ".join(row[i].ljust(widths[i]) for i in range(len(row))))


# 4.1 - время решения одной системы разными методами
def experiment_single_system(sizes, seed=42):
    print("\n=== 4.1 Сравнение времени решения одной системы ===")
    headers = ["n", "Gauss", "Gauss pivot", "LU decomp", "LU solve",
               "LU total", "Residual (pivot)", "Residual (LU)"]
    rows = []
    rng = random.Random(seed)

    for n in sizes:
        print(f"Размер n = {n}")
        a = random_matrix(n, rng)
        b = random_vector(n, rng)

        try:
            _, t_gauss = measure_time(gaussian_no_pivot, a, b)
            t_gauss = f"{t_gauss:.6f}"
        except ValueError as error:
            t_gauss = f"error: {error}"

        x_pivot, t_pivot = measure_time(gaussian_partial_pivot, a, b)
        (l, u), t_lu_decomp = measure_time(lu_decomposition, a)
        x_lu, t_lu_solve = measure_time(solve_with_lu, l, u, b)

        rows.append([
            n,
            t_gauss,
            f"{t_pivot:.6f}",
            f"{t_lu_decomp:.6f}",
            f"{t_lu_solve:.6f}",
            f"{t_lu_decomp + t_lu_solve:.6f}",
            f"{residual_norm(a, x_pivot, b):.3e}",
            f"{residual_norm(a, x_lu, b):.3e}",
        ])

    print_table(headers, rows)


# 4.2 - выгода LU при нескольких правых частях
def experiment_multiple_rhs(n, k_values, seed=42):
    print("\n=== 4.2 Экономия времени при множественных правых частях ===")
    headers = ["k", "Gauss pivot total", "LU decomp", "LU all solves", "LU total"]
    rows = []
    rng = random.Random(seed)
    a = random_matrix(n, rng)

    (l, u), t_lu_decomp = measure_time(lu_decomposition, a)

    for k in k_values:
        print(f"Количество правых частей k = {k}")
        right_parts = [random_vector(n, rng) for _ in range(k)]

        start = time.perf_counter()
        for b in right_parts:
            gaussian_partial_pivot(a, b)
        t_gauss_total = time.perf_counter() - start

        start = time.perf_counter()
        for b in right_parts:
            solve_with_lu(l, u, b)
        t_lu_solves = time.perf_counter() - start

        rows.append([
            k,
            f"{t_gauss_total:.6f}",
            f"{t_lu_decomp:.6f}",
            f"{t_lu_solves:.6f}",
            f"{t_lu_decomp + t_lu_solves:.6f}",
        ])

    print_table(headers, rows)


# 4.3 - точность на матрицах Гильберта
def experiment_hilbert(sizes):
    print("\n=== 4.3 Проверка точности на матрицах Гильберта ===")
    headers = ["n", "Method", "Relative error", "Residual"]
    rows = []

    methods = [
        ("Gauss", gaussian_no_pivot),
        ("Gauss pivot", gaussian_partial_pivot),
    ]

    for n in sizes:
        h = hilbert_matrix(n)
        x_exact = [1.0] * n
        b = mat_vec(h, x_exact)

        for name, method in methods:
            try:
                x_approx = method(h, b)
                rows.append([n, name,
                             f"{relative_error(x_approx, x_exact):.3e}",
                             f"{residual_norm(h, x_approx, b):.3e}"])
            except ValueError as error:
                rows.append([n, name, f"error: {error}", "-"])

        try:
            l, u = lu_decomposition(h)
            x_lu = solve_with_lu(l, u, b)
            rows.append([n, "LU",
                         f"{relative_error(x_lu, x_exact):.3e}",
                         f"{residual_norm(h, x_lu, b):.3e}"])
        except ValueError as error:
            rows.append([n, "LU", f"error: {error}", "-"])

    print_table(headers, rows)

File: experiments.c
#include <stdio.h>
#include <stdlib.h>
#include "matrices.h"
#include "metrics.h"
#include "solvers.h"
#include "timing.h"
#include "experiments.h"

/* 4.1 - время решения одной системы разными методами */
void experiment_single_system(const int *sizes, int count) {
    printf("\n=== 4.1 Сравнение времени решения одной системы ===\n");
    printf("%-5s %-11s %-11s %-11s %-11s %-11s %-13s %-13s\n",
           "n", "Gauss", "Gauss piv", "LU decomp", "LU solve",
           "LU total", "Res (pivot)", "Res (LU)");

    for (int s = 0; s < count; s++) {
        int n = sizes[s];
        double **a = random_matrix(n);
        double *b = random_vector(n);
        double *x = alloc_vector(n);
        double t0;

        t0 = now_seconds();
        int ok_gauss = gaussian_no_pivot(a, b, n, x);
        double t_gauss = now_seconds() - t0;

        t0 = now_seconds();
        gaussian_partial_pivot(a, b, n, x);
        double t_pivot = now_seconds() - t0;
        double res_pivot = residual_norm(a, x, b, n);   /* x сейчас от метода с выбором */

        double **l = alloc_matrix(n);
        double **u = alloc_matrix(n);

        t0 = now_seconds();
        lu_decomposition(a, n, l, u);
        double t_decomp = now_seconds() - t0;

        t0 = now_seconds();
        solve_with_lu(l, u, b, n, x);
        double t_solve = now_seconds() - t0;
        double res_lu = residual_norm(a, x, b, n);

        printf("%-5d ", n);
        if (ok_gauss == 0)
            printf("%-11.6f ", t_gauss);
        else
            printf("%-11s ", "error");
        printf("%-11.6f %-11.6f %-11.6f %-11.6f %-13.3e %-13.3e\n",
               t_pivot, t_decomp, t_solve, t_decomp + t_solve, res_pivot, res_lu);

        free_matrix(a, n);
        free_matrix(l, n);
        free_matrix(u, n);
        free(b);
        free(x);
    }
}

/* 4.2 - выгода LU при нескольких правых частях */
void experiment_multiple_rhs(int n, const int *k_values, int count) {
    printf("\n=== 4.2 Экономия времени при множественных правых частях ===\n");
    printf("%-6s %-18s %-11s %-15s %-11s\n",
           "k", "Gauss pivot total", "LU decomp", "LU all solves", "LU total");

    double **a = random_matrix(n);
    double **l = alloc_matrix(n);
    double **u = alloc_matrix(n);
    double *x = alloc_vector(n);

    double t0 = now_seconds();
    lu_decomposition(a, n, l, u);
    double t_decomp = now_seconds() - t0;

    for (int s = 0; s < count; s++) {
        int k = k_values[s];

        /* k случайных правых частей */
        double **rhs = malloc(k * sizeof(double *));
        for (int i = 0; i < k; i++)
            rhs[i] = random_vector(n);

        /* Гаусс: каждую систему решаем заново целиком */
        t0 = now_seconds();
        for (int i = 0; i < k; i++)
            gaussian_partial_pivot(a, rhs[i], n, x);
        double t_gauss = now_seconds() - t0;

        /* LU: разложение уже готово, для каждой части - только подстановки */
        t0 = now_seconds();
        for (int i = 0; i < k; i++)
            solve_with_lu(l, u, rhs[i], n, x);
        double t_lu = now_seconds() - t0;

        printf("%-6d %-18.6f %-11.6f %-15.6f %-11.6f\n",
               k, t_gauss, t_decomp, t_lu, t_decomp + t_lu);

        for (int i = 0; i < k; i++)
            free(rhs[i]);
        free(rhs);
    }

    free_matrix(a, n);
    free_matrix(l, n);
    free_matrix(u, n);
    free(x);
}

/* печать одной строки результатов для эксперимента 4.3 */
static void print_hilbert_row(int n, const char *method, int ok,
                              double rel_err, double res) {
    if (ok == 0)
        printf("%-4d %-13s %-13.3e %-13.3e\n", n, method, rel_err, res);
    else
        printf("%-4d %-13s %-13s %-13s\n", n, method, "error", "-");
}

/* 4.3 - точность на матрицах Гильберта */
void experiment_hilbert(const int *sizes, int count) {
    printf("\n=== 4.3 Проверка точности на матрицах Гильберта ===\n");
    printf("%-4s %-13s %-13s %-13s\n", "n", "Method", "Rel error", "Residual");

    for (int s = 0; s < count; s++) {
        int n = sizes[s];
        double **h = hilbert_matrix(n);
        double *x_exact = alloc_vector(n);
        for (int i = 0; i < n; i++)
            x_exact[i] = 1.0;
        double *b = mat_vec(h, x_exact, n);    /* b = H * x_exact */
        double *x = alloc_vector(n);

        int ok = gaussian_no_pivot(h, b, n, x);
        print_hilbert_row(n, "Gauss", ok,
                          relative_error(x, x_exact, n), residual_norm(h, x, b, n));

        ok = gaussian_partial_pivot(h, b, n, x);
        print_hilbert_row(n, "Gauss pivot", ok,
                          relative_error(x, x_exact, n), residual_norm(h, x, b, n));

        double **l = alloc_matrix(n);
        double **u = alloc_matrix(n);
        ok = lu_decomposition(h, n, l, u);
        if (ok == 0)
            ok = solve_with_lu(l, u, b, n, x);
        print_hilbert_row(n, "LU", ok,
                          relative_error(x, x_exact, n), residual_norm(h, x, b, n));

        free_matrix(l, n);
        free_matrix(u, n);
        free_matrix(h, n);
        free(x_exact);
        free(b);
        free(x);
    }
}
